File: benchmarks/stress_benchmarks.zig
const std = @import("std");
const dom = @import("dom");
const Document = dom.Document;
const Element = dom.Element;
const Node = dom.Node;

// ============================================================================
// STRESS TEST BENCHMARKS (Expensive Operations)
// ============================================================================

pub fn benchCreate10kNodes(allocator: std.mem.Allocator) !void {
    const doc = try Document.init(allocator);
    defer doc.release();

    const root = try Element.create(allocator, "div");
    _ = try doc.node.appendChild(root);

    // Create 10,000 elements
    var i: usize = 0;
    while (i < 10000) : (i += 1) {
        const child = try Element.create(allocator, "div");
        _ = try root.appendChild(child);
    }
}

// Note: Create+destroy test temporarily disabled due to memory management complexity
// The issue is with managing ref counts during mass removal operations
// Individual create/destroy operations work fine (see removeChild benchmark above)

pub fn benchDeepTree500Levels(allocator: std.mem.Allocator) !void {
    const doc = try Document.init(allocator);
    defer doc.release();

    const root = try Element.create(allocator, "div");
    _ = try doc.node.appendChild(root);

    // Build extremely deep tree: 500 levels
    var parent = root;
    var level: usize = 0;
    while (level < 500) : (level += 1) {
        const child = try Element.create(allocator, "div");
        _ = try parent.appendChild(child);
        parent = child;
    }

    // Traverse back up
    var count: usize = 0;
    var node: ?*Node = parent;
    while (node) |n| {
        count += 1;
        node = n.parent_node;
    }
    if (count == 0) unreachable;
}

pub fn benchComplexQuery10kNodes(allocator: std.mem.Allocator) !void {
    const doc = try Document.init(allocator);
    defer doc.release();

    const root = try Element.create(allocator, "div");
    _ = try doc.node.appendChild(root);

    // Create 10,000 elements with classes
    var i: usize = 0;
    while (i < 10000) : (i += 1) {
        const child = try Element.create(allocator, "div");

        // Every 10th element gets a special class
        if (i % 10 == 0) {
            try Element.setClassName(child, "special");
        } else {
            try Element.setClassName(child, "normal");
        }

        _ = try root.appendChild(child);
    }

    // Run complex query over 10k nodes
    const results = try Element.querySelectorAll(root, "div.special");
    defer results.deinit();
}

pub fn benchMassiveAttributeOps(allocator: std.mem.Allocator) !void {
    const doc = try Document.init(allocator);
    defer doc.release();

    const root = try Element.create(allocator, "div");
    _ = try doc.node.appendChild(root);

    // Create 1,000 elements and set 10 attributes each
    var i: usize = 0;
    while (i < 1000) : (i += 1) {
        const child = try Element.create(allocator, "div");
        _ = try root.appendChild(child);

        // Set 10 attributes per element
        try Element.setAttribute(child, "id", "element");
        try Element.setAttribute(child, "class", "container");
        try Element.setAttribute(child, "data-index", "123");
        try Element.setAttribute(child, "data-value", "test");
        try Element.setAttribute(child, "aria-label", "Label");
        try Element.setAttribute(child, "role", "button");
        try Element.setAttribute(child, "tabindex", "0");
        try Element.setAttribute(child, "title", "Title");
        try Element.setAttribute(child, "data-foo", "bar");
        try Element.setAttribute(child, "data-baz", "qux");
    }
}

pub fn benchWideTree100x100(allocator: std.mem.Allocator) !void {
    const doc = try Document.init(allocator);
    defer doc.release();

    const root = try Element.create(allocator, "div");
    _ = try doc.node.appendChild(root);

    // Create 100 parent elements, each with 100 children (10,000 total)
    var parent_idx: usize = 0;
    while (parent_idx < 100) : (parent_idx += 1) {
        const parent = try Element.create(allocator, "div");
        _ = try root.appendChild(parent);

        var child_idx: usize = 0;
        while (child_idx < 100) : (child_idx += 1) {
            const child = try Element.create(allocator, "span");
            _ = try parent.appendChild(child);
        }
    }

    // Traverse entire tree
    var count: usize = 0;
    var node = root.firstChild();
    while (node) |n| {
        count += 1;
        var child = n.firstChild();
        while (child) |c| {
            count += 1;
            child = c.nextSibling();
        }
        node = n.nextSibling();
    }
    if (count == 0) unreachable;
}

File: js-benchmarks/stress-benchmarks.js
/**
 * Stress Test Benchmarks
 * These tests involve 1,000-10,000 operations per iteration
 */

async function runStressBenchmarks() {
  await runBenchmarkSuite('STRESS TESTS (Expensive Operations)', [
    {
      name: 'Create 10,000 nodes',
      iterations: 10,
      fn: benchCreate10kNodes
    },
    {
      name: 'Deep tree (500 levels)',
      iterations: 100,
      fn: benchDeepTree500Levels
    },
    {
      name: 'Complex query over 10k nodes',
      iterations: 10,
      fn: benchComplexQuery10kNodes
    },
    {
      name: '1,000 elements × 10 attributes',
      iterations: 100,
      fn: benchMassiveAttributeOps
    },
    {
      name: 'Wide tree (100×100 = 10k nodes)',
      iterations: 10,
      fn: benchWideTree100x100
    }
  ]);
}

function benchCreate10kNodes() {
  const nodes = [];
  for (let i = 0; i < 10000; i++) {
    const elem = document.createElement('div');
    elem.className = 'item';
    nodes.push(elem);
  }
}

function benchDeepTree500Levels() {
  const root = document.createElement('div');
  let current = root;
  
  for (let i = 0; i < 500; i++) {
    const child = document.createElement('div');
    current.appendChild(child);
    current = child;
  }
}

function benchComplexQuery10kNodes() {
  // Create a large tree with 10k nodes
  const root = document.createElement('div');
  root.id = 'root';
  
  for (let i = 0; i < 100; i++) {
    const section = document.createElement('section');
    section.className = 'section';
    
    for (let j = 0; j < 100; j++) {
      const article = document.createElement('article');
      article.className = 'post';
      
      const p = document.createElement('p');
      p.className = 'content';
      article.appendChild(p);
      
      section.appendChild(article);
    }
    
    root.appendChild(section);
  }
  
  // Complex query
  const results = root.querySelectorAll('section.section > article.post > p.content');
}

function benchMassiveAttributeOps() {
  const elements = [];
  
  // Create 1,000 elements
  for (let i = 0; i < 1000; i++) {
    const elem = document.createElement('div');
    elements.push(elem);
  }
  
  // Set 10 attributes on each
  for (const elem of elements) {
    for (let i = 0; i < 10; i++) {
      elem.setAttribute(`data-attr${i}`, `value${i}`);
    }
  }
}

function benchWideTree100x100() {
  const root = document.createElement('div');
  
  // Create 100 sections, each with 100 children
  for (let i = 0; i < 100; i++) {
    const section = document.createElement('section');
    
    for (let j = 0; j < 100; j++) {
      const div = document.createElement('div');
      div.className = 'item';
      section.appendChild(div);
    }
    
    root.appendChild(section);
  }
}
